translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,388 @@
+		LabelValues:    labelValuesToOC(point.Labels, labelKeys),
+		Points: []*ocmetrics.Point{
+			{
+				Timestamp: unixnanoToTimestamp(point.TimestampUnixnano),
+				Value: &ocmetrics.Point_DoubleValue{
+					DoubleValue: point.Value,
+				},
+			},
+		},
+	}
+}
+
+func histogramPointToOC(point *otlpmetrics.HistogramDataPoint, labelKeys *labelKeys) *ocmetrics.TimeSeries {
+	return &ocmetrics.TimeSeries{
+		StartTimestamp: unixnanoToTimestamp(point.StartTimeUnixnano),
+		LabelValues:    labelValuesToOC(point.Labels, labelKeys),
+		Points: []*ocmetrics.Point{
+			{
+				Timestamp: unixnanoToTimestamp(point.TimestampUnixnano),
+				Value: &ocmetrics.Point_DistributionValue{
+					DistributionValue: &ocmetrics.DistributionValue{
+						Count:                 int64(point.Count),
+						Sum:                   point.Sum,
+						SumOfSquaredDeviation: 0,
+						BucketOptions:         histogramExplicitBoundsToOC(point.ExplicitBounds),
+						Buckets:               histogramBucketsToOC(point.Buckets),
+					},
+				},
+			},
+		},
+	}
+}
+
+func histogramExplicitBoundsToOC(bounds []float64) *ocmetrics.DistributionValue_BucketOptions {
+	if len(bounds) == 0 {
+		return nil
+	}
+
+	return &ocmetrics.DistributionValue_BucketOptions{
+		Type: &ocmetrics.DistributionValue_BucketOptions_Explicit_{
+			Explicit: &ocmetrics.DistributionValue_BucketOptions_Explicit{
+				Bounds: bounds,
+			},
+		},
+	}
+}
+
+func histogramBucketsToOC(buckets []*otlpmetrics.HistogramDataPoint_Bucket) []*ocmetrics.DistributionValue_Bucket {
+	if len(buckets) == 0 {
+		return nil
+	}
+
+	ocBuckets := make([]*ocmetrics.DistributionValue_Bucket, 0, len(buckets))
+	for _, bucket := range buckets {
+		ocBuckets = append(ocBuckets, &ocmetrics.DistributionValue_Bucket{
+			Count:    int64(bucket.Count),
+			Exemplar: exemplarToOC(bucket.Exemplar),
+		})
+	}
+	return ocBuckets
+}
+
+func exemplarToOC(exemplar *otlpmetrics.HistogramDataPoint_Bucket_Exemplar) *ocmetrics.DistributionValue_Exemplar {
+	if exemplar == nil {
+		return nil
+	}
+
+	return &ocmetrics.DistributionValue_Exemplar{
+		Value:       exemplar.Value,
+		Timestamp:   unixnanoToTimestamp(exemplar.TimestampUnixnano),
+		Attachments: exemplarAttachmentsToOC(exemplar.Attachments),
+	}
+}
+
+func exemplarAttachmentsToOC(attachments []*otlpcommon.StringKeyValue) map[string]string {
+	if len(attachments) == 0 {
+		return nil
+	}
+
+	ocAttachments := make(map[string]string, len(attachments))
+	for _, att := range attachments {
+		ocAttachments[att.Key] = att.Value
+	}
+	return ocAttachments
+}
+
+func labelValuesToOC(labels []*otlpcommon.StringKeyValue, labelKeys *labelKeys) []*ocmetrics.LabelValue {
+	if len(labels) == 0 {
+		return nil
+	}
+
+	// Initialize label values with defaults
+	// (The order matches key indices)
+	labelValues := make([]*ocmetrics.LabelValue, len(labelKeys.keyIndices))
+	for i := 0; i < len(labelKeys.keys); i++ {
+		labelValues[i] = &ocmetrics.LabelValue{
```

Comment:
This constructor can be replaced with allocating a slice `[]ocmetrics.LabelValue` and setting pointers, since `HasValue: false` is a zero value anyway.
But explicit code is more readable in case we need to customize the defaults.